// @temporalio/worker/lib/workflow/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initLoggerSink = void 0;
const runtime_1 = require("../runtime");
/**
 * Injects a logger sink that forwards to the worker's logger
 */
function initLoggerSink(logger) {
    return {
        __temporal_logger: {
            trace: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.trace(message, attrs);
                },
            },
            debug: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.debug(message, attrs);
                },
            },
            info: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.info(message, attrs);
                },
            },
            warn: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.warn(message, attrs);
                },
            },
            error: {
                fn(_, message, attrs) {
                    logger ?? (logger = runtime_1.Runtime.instance().logger);
                    logger.error(message, attrs);
                },
            },
        },
    };
}
exports.initLoggerSink = initLoggerSink;
