// @temporalio/interceptors-opentelemetry/lib/workflow/definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SPAN_DELIMITER = exports.SpanName = void 0;
var SpanName;
(function (SpanName) {
    /**
     * Workflow is scheduled by a client
     */
    SpanName["WORKFLOW_START"] = "StartWorkflow";
    /**
     * Workflow is client calls signalWithStart
     */
    SpanName["WORKFLOW_SIGNAL_WITH_START"] = "SignalWithStartWorkflow";
    /**
     * Workflow run is executing
     */
    SpanName["WORKFLOW_EXECUTE"] = "RunWorkflow";
    /**
     * Child Workflow is started (by parent Workflow)
     */
    SpanName["CHILD_WORKFLOW_START"] = "StartChildWorkflow";
    /**
     * Activity is scheduled by a Workflow
     */
    SpanName["ACTIVITY_START"] = "StartActivity";
    /**
     * Activity is executing
     */
    SpanName["ACTIVITY_EXECUTE"] = "RunActivity";
    /**
     * Workflow is continuing as new
     */
    SpanName["CONTINUE_AS_NEW"] = "ContinueAsNew";
})(SpanName || (exports.SpanName = SpanName = {}));
exports.SPAN_DELIMITER = ':';
